File: python/Prog_주사위고르기.py
from itertools import combinations

def brute_x(x, idx, total, x_arr) :
    if idx == len(x) :
        x_arr.append(total)
        return

    for a in x[idx] :
        total += a
        brute_x(x, idx+1, total, x_arr)
        total -= a
        
def binary_search(x, arr) :
    lo = 0
    hi = len(arr)-1
    
    while lo <= hi :
        mid = (lo + hi)//2
        if arr[mid] <= x : lo = mid+1
        else : hi = mid-1
    return len(arr) - lo

def brute_y(y, idx, total, x_arr, res) :
    if idx == len(y) :
        res[0] += binary_search(total, x_arr)
        return
    for a in y[idx] :
        total += a
        brute_y(y, idx+1, total, x_arr, res)
        total -= a
    

def solution(dice):
    arr = [i for i in range(len(dice))]
    
    ans = []
    for comb in combinations(arr, len(dice)//2) :
        x = []
        y = []
        
        for i in range(len(dice)) :
            if i in comb : x.append(dice[i])
            else : y.append(dice[i])
        
        x_arr = []
        brute_x(x, 0, 0, x_arr)
        x_arr.sort()
        
        res = [0]
        brute_y(y, 0, 0, x_arr, res)
        
        ans.append([res[0], [i + 1 for i in comb]])
        
    ans.sort(key = lambda x : x[0])
    return ans[-1][1]
        
        
                
        

File: java/src/Prog_주사위고르기.java
import java.util.*;

class Solution {
    public int[] solution(int[][] dice) {
        int N = dice.length;
        
        Map<List<Integer>, List<Integer>> results = new HashMap<>();
        
        int[] idxes = new int[N];
        
        for (int i=0;i<N;i++) idxes[i] = i;
        
        List<List<Integer>> combinations = new ArrayList<>();
        combine(idxes, N/2, 0, new ArrayList<>(), combinations);
        
        for (List<Integer> combination : combinations) {
            List<Integer> result = new ArrayList<>();
            getResult(combination, result, 0, 0, dice);
            Collections.sort(result);
            results.put(combination, result);
        }
        
        
        int ans[] = new int[N/2];
        
        int maxCount = 0;
        
        for (List<Integer> combination : combinations) {
            List<Integer> another = getAnother(combination, N);
            
            List<Integer> result_cur = results.get(combination);
            List<Integer> result_another = results.get(another);
            
            int tmp = 0;
            for (int x : result_cur) {
                tmp += lowerBound(result_another, x);
            }
            
            if (maxCount < tmp) {
                for (int i=0;i<N/2;i++) {
                    ans[i] = combination.get(i)+1;
                }
                maxCount = tmp;
            }
        }
        
        return ans;
    }
    
    public int lowerBound(List<Integer> result, int x) {
        int lo = 0;
        int hi = result.size()-1;
        
        while (lo <= hi) {
            int mid = (lo+hi)/2;
            
            if (result.get(mid) < x) lo = mid+1;
            else hi = mid-1;
        }
        
        return lo;
    }
    
    public List<Integer> getAnother(List<Integer> combination, int N) {
        boolean[] selected = new boolean[N];
        for (int n : combination) {
            selected[n] = true;
        }
        
        List<Integer> ret = new ArrayList<>();
        
        for (int i=0;i<N;i++) {
            if (!selected[i]) ret.add(i);
        }
        return ret;
    }
    
    public void getResult(List<Integer> combination, List<Integer> result, int idx, int sum, int[][] dice) {
        if (idx == combination.size()) {
            result.add(sum);
            return;
        }
        
        for (int i=0;i<6;i++) {
            getResult(combination, result, idx+1, sum + dice[combination.get(idx)][i], dice);
        }
    }
    
    public void combine(int[] idxes, int r, int start, List<Integer> current, List<List<Integer>> combinations) {
        if (current.size() == r) {
            combinations.add(new ArrayList<>(current));
            return;
        }
        
        for (int i=start;i<idxes.length;i++) {
            current.add(idxes[i]);
            combine(idxes, r, i+1, current, combinations);
            current.remove(current.size()-1);
        }
    }
}